components/builder-db/src/pool.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2016 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::ops::{Deref, DerefMut};
+use std::result;
+use std::sync::atomic::{AtomicUsize, Ordering, ATOMIC_USIZE_INIT};
+use std::thread;
+use std::time::Duration;
+
+use r2d2;
+use r2d2_postgres::{self, PostgresConnectionManager, TlsMode};
+use postgres::transaction::Transaction;
+use postgres;
+
+use error::{Error, Result};
+
+// We will use this to allocate test schmeas
+static GLOBAL_SCHEMA_COUNT: AtomicUsize = ATOMIC_USIZE_INIT;
+
+#[derive(Debug, Clone)]
+pub struct Pool {
+    inner: r2d2::Pool<PostgresConnectionManager>,
+}
+
+impl Pool {
+    pub fn new(connection_url: &str,
```

Comment:
This is actually not easy to do, I take it back. We would need to wrap the inner section in an Option to make this work, since we need the pool to be persistent, and the pool object itself will sleep. We could move the retry logic out if we wanted, but you're still going to sleep the thread. Ie: you wouldn't ever call `Pool.new()` anymore, because it literally can't return from an error - so you'll call `Pool.start()` all the time, which I'm not sure is better.